resqpy/grid_surface.py

```diff
@@ -1494,6 +1497,351 @@ def find_faces_to_represent_surface_regular(grid,
+        return_normal_vectors (bool): if True, an array of normal vectors is created for the gcs,
+            holding a unit vector normal to the surface at the centre of each face; it is returned
+            with the gcs.
+        normals (np.ndarray): array of normal vectors to the surface at the centre of its
+            corresponding cell face.
+        cwt (np.ndarray): array of all the triangle indices creating each triangle in clockwise form.
+
+    Returns:
+        Tuple containing:
+
+        - faces (np.ndarray): boolean array of each cell face that can represent the surface.
+        - sides (np.ndarray): array of the sides creating the cells in the grid.
+        - normals (np.ndarray): array of normal vectors to the surface at the centre of its
+            corresponding cell face.
+    """
+    hit_stack = np.stack(where_true(hits), axis = -1)
+    for i in numba.prange(len(hit_stack)):
+        hit = hit_stack[i]
+        tri = hit[0]
+
+        # Two dimensions for the centre point in this projection.
+        d1 = hit[1]
+        d2 = hit[2]
+
+        ind0 = np.zeros(3, dtype = numba.int32)
+        ind0[index1] = d1
+        ind0[index2] = d2
+
+        ind1 = np.zeros(3, dtype = numba.int32)
+        ind1[index1] = d1
+        ind1[index2] = d2
+        ind1[2 - axis] = -1
+
+        xyz = meet.line_triangle_intersect_numba(
+            centres[ind0[0], ind0[1], ind0[2]],
+            centres[ind1[0], ind1[1], ind1[2]] - centres[ind0[0], ind0[1], ind0[2]],
+            points[triangles[tri]],
+            line_segment = True,
+            t_tol = 1.0e-6,
+        )
+        if xyz is None:  # meeting point is outwith grid
+            continue
+
+        ind2 = np.zeros(4, dtype = np.int32)
+        ind2[index1] = d1
+        ind2[index2] = d2
+        ind2[-1] = axis
+
+        face = int((xyz[axis] - centres[ind2[0], ind2[1], ind2[2], ind2[3]]) / grid_dxyz[axis])
+        ind_face = np.zeros(3, dtype = np.int32)
+        ind_face[index1] = d1
+        ind_face[index2] = d2
+        ind_face[2 - axis] = face
+
+        faces[ind_face[0], ind_face[1], ind_face[2]] = True
+        if consistent_side:
+            sides[ind_face[0], ind_face[1], ind_face[2]] = cwt[tri]
+        if return_normal_vectors:
+            normals[ind_face[0], ind_face[1], ind_face[2]] = vec.triangle_normal_vector_numba(points[triangles[tri]])
+            # todo: if consistent side, could deliver information about horizon surface inversion
+            if normals[ind2[0], ind2[1], ind2[2], ind2[3]] > 0.0:
+                normals[ind_face[0], ind_face[1], ind_face[2]] = -normals[ind_face[0], ind_face[1], ind_face[2]]
+    return faces, sides, normals
+
+
+def find_faces_to_represent_surface_regular_optimised(
+    grid,
+    surface,
+    name,
+    title = None,
+    centres = None,
+    progress_fn = None,
+    consistent_side = False,
+    return_normal_vectors = False,
```

Comment:
Please can we make this compatible with the latest _regular() version?  (Will require manual merging of the offset related code.)

Reply:
Yeah I've just added that